src/tailscale/cli.ts

```diff
@@ -100,42 +100,109 @@ export class Tailscale {
-
-        this.childProcess.stderr.on('end', () => {
-          // Process the remaining data in the buffer after the stream ends
-          const line = buffer.trim();
+        // Process all complete lines except the last one
+        for (let i = 0; i < lines.length - 1; i++) {
+          const line = lines[i].trim();
           if (line.length > 0) {
             Logger.info(line, LOG_COMPONENT);
           }
-        });
-      } else {
-        Logger.error('childProcess.stderr is null', LOG_COMPONENT);
-        throw new Error('childProcess.stderr is null');
+        }
+
+        buffer = lines[lines.length - 1];
+      });
+
+      childProcess.stderr.on('end', () => {
+        // Process the remaining data in the buffer after the stream ends
+        const line = buffer.trim();
+        if (line.length > 0) {
+          Logger.info(line, LOG_COMPONENT);
+        }
+      });
+    } else {
+      Logger.error('childProcess.stderr is null', LOG_COMPONENT);
+      throw new Error('childProcess.stderr is null');
+    }
+  }
+
+  async initSudo() {
+    return new Promise<null>((resolve, err) => {
+      const binPath = this.tsrelayPath();
+      const args = [`-nonce=${this.nonce}`, `-port=${this.port}`];
+      if (this._vscode.env.logLevel === LogLevel.Debug) {
+        args.push('-v');
       }
+      Logger.info(`path: ${binPath}`, LOG_COMPONENT);
+      this.notifyExit = () => {
+        Logger.info('starting sudo tsrelay');
+        const childProcess = cp.spawn(`/usr/bin/pkexec`, [
+          '--disable-internal-agent',
+          binPath,
+          ...args,
+        ]);
+        childProcess.on('exit', async (code) => {
+          Logger.warn(`sudo child process exited with code ${code}`, LOG_COMPONENT);
+          if (code === 0) {
+            return;
+          } else if (code === 126) {
+            // authentication not successful
+            this._vscode.window.showErrorMessage(
+              'Creating funnels must only be done by an administrator'
+            );
+          } else {
+            this._vscode.window.showErrorMessage('Could not run authenticator, please check logs');
+          }
+          await this.init(this.port, this.nonce);
+          err('unauthenticated');
+        });
+        childProcess.on('error', (err) => {
+          Logger.error(`sudo child process error ${err}`, LOG_COMPONENT);
+        });
+        childProcess.stdout.on('data', (data: Buffer) => {
+          Logger.debug('received data from sudo');
+          const details = JSON.parse(data.toString().trim()) as TSRelayDetails;
+          if (this.url !== details.address) {
+            Logger.error(`expected url to be ${this.url} but got ${details.address}`);
+            return;
+          }
+          this.runPortDisco();
+          Logger.debug('resolving');
+          resolve(null);
+        });
+        this.processStderr(childProcess);
+      };
+      Logger.info('shutting down tsrelay');
+      this.childProcess!.kill('SIGINT');
```

Comment:
Also remove the non-null assertion here. Maybe just call dispose?

Reply:
Will wrap in an if statement